### PR TITLE
Keyword Matching

### PR DESCRIPTION
**Pull request by ivanleomk**
This PR introduces a new flag in the `evaluate` command called `eval-mode`. This helps to support keyword matching where we try to match rows that have a text content which has at least 3 rows. 

You can run the text search by doing 

```
rag-app evaluate from-jsonl --input-file-path ./output.jsonl --db-path ./db --table-name pg --eval-mode fts
```

When we generate bad keywords using a llm, then we get much fewer rows ( Less than 25 ) since our query only returns rows that have at least one match. This results in some queries returning 0 or 1 rows which results in a 'n/a' or 0/1 value for the `ndcg` respectively.

I'm not sure if this is the behaviour that we might want so will explore other queries. 

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 184dcaf9c9f424c195672d97aa51ca0cbd156029.  | 
|--------|--------|

### Summary:
This PR introduces a new `fts` mode to the `evaluate` command in `rag_app` for keyword matching, updates the `output.jsonl` file, and modifies the `calculate_ndcg` function in `rag_app/src/metrics.py` to handle cases with 0 or 1 predictions.

**Key points**:
- Introduced a new flag `eval-mode` in the `evaluate` command of `rag_app` to support keyword matching.
- Added `fts` (Full Text Search) mode that generates keywords for questions and matches chunks with these keywords.
- Updated `output.jsonl` file.
- Modified `calculate_ndcg` function in `rag_app/src/metrics.py` to handle cases with 0 or 1 predictions.
- Added uncertainty about the behavior of the new mode and plans to explore other queries.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: rag_app/evaluate.py [hunk 1]
```diff
@@ -83,12 +93,93 @@ async def query_table(query: EmbeddedEvaluationItem):
...
+        keywords = query.keywords
+        keywords = [keyword.replace("'", "''") for keyword in keywords]
+        keyword_conditions = " +\n".join(
+            [
+                f"CASE WHEN regexp_matches(text, '(?i){keyword}') THEN 1 ELSE 0 END"
+                for keyword in keywords
+            ]
+        )
+        keyword_search = "|".join(keywords)
+        duckdb_query = f"""
```
[the remark this answers is not in the document]
Comment (ivanleomk) [reply rag_app/evaluate.py:146]: @ellipsis-dev Fix this issue using a duck-db specific method

**Comment by ivanleomk** [comment]
Seems like the python lib also supports FTS using an index through tantivity (https://lancedb.github.io/lancedb/fts/#index-multiple-columns )

Will look into this

**Comment by ivanleomk** [comment]
Current FTS results

```
                                  MRR@3  MRR@5  MRR@10  MRR@20 NDCG@3 NDCG@5 NDCG@10 NDCG@20  retrieved_size
chunk_id                                                                                                    
09369eb77f4c743034d01d13744faf6d    1.0    1.0     1.0    1.00    1.0    1.0     1.0     1.0               7
b6bb3ecf14a22bc3144b3c8bb101a1e8    1.0    1.0     1.0    1.00    1.0    1.0     1.0     1.0              25
65d0dc53f68d4e5a9b95e6268673cc09    1.0    1.0     1.0    1.00    1.0    1.0     1.0     1.0              25
f8270d09dace076cb5dbd79309d08fa4    0.0    0.0     0.0    0.05    0.0    0.0     0.0    0.23              25
de494946e19340c348b991e5845cd8c4    1.0    1.0     1.0    1.00    1.0    1.0     1.0     1.0               4
afc4b95c2537f788fe5711c9835b58bf    0.0    0.0     0.0    0.00      0      0       0       0               1
66288d8fe7e8f36b7f4c2bf4d5af7b18    1.0    1.0     1.0    1.00    1.0    1.0     1.0     1.0               9
3bb3e61a043cd396acc7669d021ab532    0.0    0.0     0.0    0.00    N/A    N/A     N/A     N/A               0
ddd2c319f6b2bae1f9583b497bc615e4    0.0    0.0     0.0    0.00      0      0       0       0               1
965e5abc2b1409f099518c51d13d7a5a    0.0    0.0     0.0    0.00      0      0       0       0               1

              Mean Values              
┏━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┓
┃ Metric         ┃ Value              ┃
┡━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━┩
│ MRR@3          │ 0.5                │
│ MRR@5          │ 0.5                │
│ MRR@10         │ 0.5                │
│ MRR@20         │ 0.505              │
│ NDCG@3         │ 0.5555555555555556 │
│ NDCG@5         │ 0.5555555555555556 │
│ NDCG@10        │ 0.5555555555555556 │
│ NDCG@20        │ 0.5811111111111111 │
│ retrieved_size │ 9.8                │
```

This is the respective semantic search result

```
                                  MRR@3  MRR@5  MRR@10  MRR@20  NDCG@3  NDCG@5  NDCG@10  NDCG@20  retrieved_size
chunk_id                                                                                                        
09369eb77f4c743034d01d13744faf6d    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
b6bb3ecf14a22bc3144b3c8bb101a1e8    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
65d0dc53f68d4e5a9b95e6268673cc09    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
f8270d09dace076cb5dbd79309d08fa4    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
de494946e19340c348b991e5845cd8c4    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
afc4b95c2537f788fe5711c9835b58bf    0.5    0.5     0.5     0.5    0.63    0.63     0.63     0.63              25
66288d8fe7e8f36b7f4c2bf4d5af7b18    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
3bb3e61a043cd396acc7669d021ab532    1.0    1.0     1.0     1.0    1.00    1.00     1.00     1.00              25
ddd2c319f6b2bae1f9583b497bc615e4    0.0    0.2     0.2     0.2    0.00    0.39     0.39     0.39              25
965e5abc2b1409f099518c51d13d7a5a    0.5    0.5     0.5     0.5    0.63    0.63     0.63     0.63              25

       Mean Values        
┏━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Metric         ┃ Value ┃
┡━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ MRR@3          │ 0.8   │
│ MRR@5          │ 0.82  │
│ MRR@10         │ 0.82  │
│ MRR@20         │ 0.82  │
│ NDCG@3         │ 0.826 │
│ NDCG@5         │ 0.865 │
│ NDCG@10        │ 0.865 │
│ NDCG@20        │ 0.865 │
│ retrieved_size │ 25.0  │
```

File: rag_app/models.py [hunk 2]
```diff
@@ -64,3 +64,12 @@ class EvaluationDataItem(BaseModel):
     answer: str
     chunk: str
     chunk_id: str
+
+
+class KeywordExtractionResponse(BaseModel):
+    keywords: List[str] = Field(
+        ...,
+        title="Keywords",
+        description="A list of at least 3 relevant keywords extracted from the question.",
```
Comment (jxnl) [inline rag_app/models.py]: can you try to improve this prompt

"say, the goal is to render a range of keywords, expand acryonms used synonoms etc etc

Comment (ivanleomk) [reply rag_app/models.py]: I played around with the prompt a bit for fts, got a small improvement

```
              Mean Values              
┏━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┓
┃ Metric         ┃ Value              ┃
┡━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━┩
│ MRR@3          │ 0.6198             │
│ MRR@5          │ 0.6328             │
│ MRR@10         │ 0.642              │
│ MRR@20         │ 0.6478             │
│ NDCG@3         │ 0.653469387755102  │
│ NDCG@5         │ 0.6781632653061224 │
│ NDCG@10        │ 0.6993877551020409 │
│ NDCG@20        │ 0.7204081632653062 │
│ retrieved_size │ 17.12              │
```

This is the semantic

```
┏━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┓
┃ Metric         ┃ Value              ┃
┡━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━┩
│ MRR@3          │ 0.8031999999999999 │
│ MRR@5          │ 0.8131999999999999 │
│ MRR@10         │ 0.8131999999999999 │
│ MRR@20         │ 0.8162             │
│ NDCG@3         │ 0.8281999999999999 │
│ NDCG@5         │ 0.8453999999999999 │
│ NDCG@10        │ 0.8453999999999999 │
│ NDCG@20        │ 0.8558             │
│ retrieved_size │ 25.0               │
```

**Comment by ivanleomk** [comment]
Implemented the Bm25 search! It's a bit slower than embedding search but performs quite well tbh

```
       Mean Values        
┏━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Metric         ┃ Value ┃
┡━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ MRR@3          │ 0.8   │
│ MRR@5          │ 0.81  │
│ MRR@10         │ 0.81  │
│ MRR@20         │ 0.82  │
│ NDCG@3         │ 0.83  │
│ NDCG@5         │ 0.85  │
│ NDCG@10        │ 0.85  │
│ NDCG@20        │ 0.86  │
│ retrieved_size │ 25.0  │
```

**Pull request merged**